pyvelox/substrait.cpp

```diff
@@ -0,0 +1,234 @@
+  PySubstraitContext::getInstance().finalize();
+}
+
+std::vector<std::shared_ptr<facebook::velox::connector::ConnectorSplit>>
+makeSplits(
+    const facebook::velox::substrait::SubstraitVeloxPlanConverter& converter,
+    std::shared_ptr<const core::PlanNode> planNode,
+    const std::string& dirPath) {
+  const auto& splitInfos = converter.splitInfos();
+  auto leafPlanNodeIds = planNode->leafPlanNodeIds();
+  // Only one leaf node is expected here.
+  EXPECT_EQ(1, leafPlanNodeIds.size());
+  const auto& splitInfo = splitInfos.at(*leafPlanNodeIds.begin());
+
+  const auto& paths = splitInfo->paths;
+  const auto& starts = splitInfo->starts;
+  const auto& lengths = splitInfo->lengths;
+  const auto fileFormat = splitInfo->format;
+
+  std::vector<std::shared_ptr<facebook::velox::connector::ConnectorSplit>>
+      splits;
+  splits.reserve(paths.size());
+
+  for (int i = 0; i < paths.size(); i++) {
+    auto path = fmt::format("{}{}", dirPath, paths[i]);
+    auto start = starts[i];
+    auto length = lengths[i];
+    auto split = facebook::velox::exec::test::HiveConnectorSplitBuilder(path)
+                     .fileFormat(fileFormat)
+                     .start(start)
+                     .length(length)
+                     .build();
+    splits.emplace_back(split);
+  }
+  return splits;
+}
+
+static inline RowVectorPtr runSubstraitQuery(
+    const std::string& plan,
+    bool enableSplits,
+    const std::string& dirPath) {
+  memory::MemoryPool* pool = PyVeloxContext::getSingletonInstance().pool();
+
+  facebook::velox::substrait::SubstraitVeloxPlanConverter planConverter(pool);
+
+  ::substrait::Plan substraitPlan;
+  // convert plan to protobuf
+  // read from file
+  if (isJSON(plan)) {
+    readFromJSON(plan, substraitPlan);
+  } else if (isFilePath(plan)) {
+    // check whether input file is in JSON format
+    if (isJsonFile(plan)) {
+      readFromFile(plan, substraitPlan);
+    } else {
+      /// other formats are not yet supported.
+      /// NOTE: To support Protobuf format, we should generate Substrait proto
+      /// in Python format too. Then we should be able to add proto format too.
+      throw py::value_error("plan should be path to a plan in JSON format.");
+    }
+  } else {
+    throw py::value_error("Invalid Substrait plan.");
+  }
+
+  auto planNode = planConverter.toVeloxPlan(substraitPlan);
+
+  if (enableSplits) {
+    return facebook::velox::exec::test::AssertQueryBuilder(planNode)
+        .splits(makeSplits(planConverter, planNode, dirPath))
+        .copyResults(pool);
+  } else {
+    return facebook::velox::exec::test::AssertQueryBuilder(planNode)
+        .copyResults(pool);
+  }
+}
+
+void addSubstraitBindings(py::module& m, bool asModuleLocalDefinitions) {
+  using namespace facebook::velox;
+  functions::prestosql::registerAllScalarFunctions();
+  aggregate::prestosql::registerAllAggregateFunctions();
+  m.def(
+      "initialize_substrait",
+      &initializeSubstrait,
```

Comment:
I think substrait should be its own module , so importing the module initializes substrait.

Reply:
Good point, probably need to see how this can be done.